irr/src/CGLTFMeshFileLoader.cpp

```diff
@@ -0,0 +1,610 @@
+		transformVertices(*vertices, parent->GlobalMatrix);
+		
+		auto maybeIndices = getIndices(meshIdx, j);
+		std::vector<u16> indices;
+		if (maybeIndices.has_value()) {
+			indices = std::move(*maybeIndices);
+			checkIndices(indices, vertices->size());
+		} else {
+			// Non-indexed geometry
+			indices = generateIndices(vertices->size());
+		}
+
+		auto *meshbuf = m_irr_model->addMeshBuffer();
+		meshbuf->append(vertices->data(), vertices->size(),
+			indices.data(), indices.size());
+	}
+}
+
+// Base transformation between left & right handed coordinate systems.
+// This just inverts the Z axis.
+static core::matrix4 leftToRight = core::matrix4(
+	1, 0, 0, 0,
+	0, 1, 0, 0,
+	0, 0, -1, 0,
+	0, 0, 0, 1
+);
+static core::matrix4 rightToLeft = leftToRight;
+
+static core::matrix4 loadTransform(const tiniergltf::Node::Matrix &m)
+{
+	// Note: Under the hood, this casts these doubles to floats.
+	return core::matrix4(
+			m[0], m[1], m[2], m[3],
+			m[4], m[5], m[6], m[7],
+			m[8], m[9], m[10], m[11],
+			m[12], m[13], m[14], m[15]);
+}
+
+static core::matrix4 loadTransform(const tiniergltf::Node::TRS &trs)
+{
+	const auto &trans = trs.translation;
+	const auto &rot = trs.rotation;
+	const auto &scale = trs.scale;
+	core::matrix4 transMat;
+	transMat.setTranslation(core::vector3df(trans[0], trans[1], trans[2]));
+	core::matrix4 rotMat = core::quaternion(rot[0], rot[1], rot[2], rot[3]).getMatrix();
+	core::matrix4 scaleMat;
+	scaleMat.setScale(core::vector3df(scale[0], scale[1], scale[2]));
+	return transMat * rotMat * scaleMat;
+}
+
+static core::matrix4 loadTransform(std::optional<std::variant<tiniergltf::Node::Matrix, tiniergltf::Node::TRS>> transform) {
+	if (!transform.has_value()) {
+		return core::matrix4();
+	}
+	core::matrix4 mat = std::visit([](const auto &t) { return loadTransform(t); }, *transform);
+	return rightToLeft * mat * leftToRight;
+}
+
+void CGLTFMeshFileLoader::MeshExtractor::loadNode(
+		const std::size_t nodeIdx,
+		ISkinnedMesh::SJoint *parent) const
+{
+	const auto &node = m_gltf_model.nodes->at(nodeIdx);
+	auto *joint = m_irr_model->addJoint(parent);
+	const core::matrix4 transform = loadTransform(node.transform);
+	joint->LocalMatrix = transform;
+	joint->GlobalMatrix = parent ? parent->GlobalMatrix * joint->LocalMatrix : joint->LocalMatrix;
+	if (node.name.has_value()) {
+		joint->Name = node.name->c_str();
+	}
+	if (node.mesh.has_value()) {
+		loadMesh(*node.mesh, joint);
+	}
+	if (node.children.has_value()) {
+		for (const auto &child : *node.children) {
+			loadNode(child, joint);
+		}
+	}
+}
+
+void CGLTFMeshFileLoader::MeshExtractor::loadNodes() const
+{
+	std::vector<bool> isChild(m_gltf_model.nodes->size());
+	for (const auto &node : *m_gltf_model.nodes) {
+		if (!node.children.has_value())
+			continue;
+		for (const auto &child : *node.children) {
+			isChild[child] = true;
+		}
+	}
+	// Load all nodes that aren't children.
+	// Children will be loaded by their parent nodes.
```

Comment:
why not merge the two loops?

Reply:
they can't be merged: the first loop determines for each node whether it is a child - that is, whether there is any node which is its parent. the second loop requires this information to be complete.

or in other words: if they were merged, the second loop could run `loadNode` for child nodes when their parents come later in the list, which it shouldn't.